src/main/java/org/squiddev/plethora/integration/astralsorcery/IntegrationAstralSorcery.java

```diff
@@ -0,0 +1,777 @@
+		@Override
+		public ItemStack getExample() {
+			ItemStack paperStack = new ItemStack(ItemsAS.constellationPaper);
+
+			IWeakConstellation constellation = getExampleConstellation();
+			if (constellation == null) return  null;
+
+			ItemConstellationPaper.setConstellation(paperStack, constellation);
+
+			return paperStack;
+		}
+	};
+
+	public static final IMetaProvider<ItemStack> META_ITEM_JOURNAL = new ItemStackContextMetaProvider<ItemJournal>(
+		ItemJournal.class,
+		"Provides the Constellation Papers stored in this Journal"
+	) {
+		@Nonnull
+		@Override
+		public Map<String, Object> getMeta(@Nonnull IPartialContext<ItemStack> context, @Nonnull ItemJournal item) {
+			ItemStack stack = context.getTarget();
+
+			//ItemJournal also has a method to map an IInventory, BUT said IInventory does not apply any
+			// of the slot restrictions imposed by the ContainerJournal...
+			//REFINE We mainly care about the name of the constellation on the paper...
+			return Collections.singletonMap("papers",
+				Arrays.stream(ItemJournal.getStoredConstellationStacks(stack))
+					.map(paper -> context.makePartialChild(paper).getMeta())
+					.collect(LuaList.toLuaList())
+					.asMap());
+		}
+
+		@Nullable
+		@Override
+		public ItemStack getExample() {
+			//FIXME Set the example
+			return null;
+		}
+	};
+
+	public static final IMetaProvider<TileWell> META_LIGHT_WELL = new BaseMetaProvider<TileWell>(
+		"Provides the catalyst item in this Lightwell"
+	) {
+
+		@Nonnull
+		@Override
+		public Map<String, Object> getMeta(@Nonnull IPartialContext<TileWell> context) {
+
+			//TODO turtle.placeDown() doesn't work to insert the catalyst; does this simulate a click, or just try to place a block?
+			// Further, the Lightwell won't accept a catalyst if a block is directly above it...
+			// And it needs a clear line to the sky... looking a bit like 'No automation via CC for you!'...
+
+			//REFINE Do we want this stack wrapped, or should we just `getMeta`?
+			return Collections.singletonMap("catalyst",
+				ContextHelpers.wrapStack(context, context.getTarget().getInventoryHandler().getStackInSlot(0)));
+		}
+
+		@Nonnull
+		@Override
+		public TileWell getExample() {
+			TileWell well = new TileWell();
+			well.getInventoryHandler().setStackInSlot(0, new ItemStack(ItemsAS.craftingComponent));
+			return well;
+		}
+	};
+
+	public static final IMetaProvider<TileCollectorCrystal> META_TILE_COLLECTOR_CRYSTAL = new BaseMetaProvider<TileCollectorCrystal>() {
+
+		@Nonnull
+		@Override
+		public Map<String, Object> getMeta(@Nonnull IPartialContext<TileCollectorCrystal> context) {
+			Map<String, Object> out = new HashMap<>(4);
+			TileCollectorCrystal target = context.getTarget();
+
+			out.putAll(context.makePartialChild(target.getCrystalProperties()).getMeta());
+
+			IWeakConstellation mainConstellation = target.getConstellation();
+			out.put("constellation", mainConstellation != null ? context.makePartialChild(mainConstellation).getMeta() : null);
```

Comment:
Maybe do `if(mainConstellation != null) out.put(...)` instead? Dunno.

Reply:
Thinking about it, CC does drop any `null` entries anyway, so that makes sense; I had been thinking more of 'show that there can/should be a value here'.  We could even do an inspection or external annotation to treat the value param of `Map.put` as `@Nonnull`, if you want.